pkg/cmd/auth-proxy/verifiers/jwt/provider.go

```
@@ -0,0 +1,111 @@
+
+	djwt "github.com/dgrijalva/jwt-go"
+	log "github.com/sirupsen/logrus"
+)
+
+type provider struct {
+	Options
+	// PublicKey is the key used to sign
+	PublicKey *rsa.PublicKey
+}
+
+// New creates and returns an oidc provider
+func New(options Options) (verifiers.Interface, error) {
+	if options.Signer == nil {
+		return nil, errors.New("no signer defined")
+	}
+
+	key, err := djwt.ParseRSAPublicKeyFromPEM(options.Signer)
+	if err != nil {
+		return nil, err
+	}
+
+	return &provider{Options: options, PublicKey: key}, nil
+}
+
+// Admit checks the token is valid
+func (o *provider) Admit(request *http.Request) (bool, error) {
+	// @step: extract the token from the request
+	bearer, found := utils.GetBearerToken(request.Header.Get("Authorization"))
+	if !found {
+		return false, errors.New("no authorization token")
+	}
+	log.Debug("checking against the openid verifier")
+
+	c := make(djwt.MapClaims)
+
+	// @step: parse and extract the identity
+	token, err := djwt.ParseWithClaims(bearer, &c, func(token *djwt.Token) (interface{}, error) {
+		return o.PublicKey, nil
+	})
+	switch err.(type) {
+	case nil:
+		if !token.Valid {
+			return false, nil
+		}
+	default:
+		return false, nil
+	}
+
+	claims := utils.NewClaims(c)
+
+	// @step: check the audience
+	if aud, found := claims.GetAudience(); !found {
+		log.Warn("no audience in the presented token")
+
+		return false, nil
+	} else if aud != "kubernetes" {
+		log.Warn("invalid audience presented in the token")
+
+		return false, nil
+	}
+
+	// @step: check the username
+	username, found := claims.GetUserClaim("preferred_username", "email")
+	if !found {
+		return false, errors.New("no username found in the identity token")
+	}
+
+	// @step: ensure no impersonation is passed through by clearing all headers
+	for name := range request.Header {
+		if strings.HasPrefix(name, "Impersonate") {
+			log.Warn("request has an impersonation header, denying the request")
+
+			return false, nil
```

maybe return an error here too? I'm not sure whether we should swallow errors or make sure it's easier to debug an issue.